src/wallet/wallet.cpp

```diff
@@ -1902,95 +1909,123 @@ void CWallet::ResendWalletTransactions(int64_t nBestBlockTime, CConnman* connman
  * @{
  */
 
+void CWallet::MarkBalancesDirty()
+{
+    fBalancesDirty = true;
+    BalancesDidChange();
+}
 
 CAmount CWallet::GetBalance() const
 {
-    CAmount nTotal = 0;
+    if (!fBalancesDirty) {
+        return nBalanceCache;
+    }
+
+    nBalanceCache = 0;
```

Comment:
I think you need to move this line after the `LOCK2(cs_main, cs_wallet);` line. Otherwise if this method is called from two different threads simultaneously, this line could zero out a balance which is in the middle of being added up by the other thread.

Or alternately, delete this line and and set nBalanceCache = nTotal atomically after the loop.